src/runtime-rs/crates/resource/src/network/network_with_netns.rs

```diff
@@ -0,0 +1,215 @@
+// Copyright (c) 2019-2022 Alibaba Cloud
+// Copyright (c) 2019-2022 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::sync::{
+    atomic::{AtomicU32, Ordering},
+    Arc,
+};
+
+use anyhow::{anyhow, Context, Result};
+use async_trait::async_trait;
+use futures::stream::TryStreamExt;
+use hypervisor::Hypervisor;
+use scopeguard::defer;
+use tokio::sync::RwLock;
+
+use super::{
+    endpoint::{Endpoint, PhysicalEndpoint, VethEndpoint},
+    network_entity::NetworkEntity,
+    network_info::network_info_from_link::NetworkInfoFromLink,
+    utils::{link, netns},
+    Network,
+};
+use crate::network::NetworkInfo;
+
+#[derive(Debug)]
+pub struct NetworkWithNetNsConfig {
+    pub network_model: String,
+    pub netns_path: String,
+    pub queues: usize,
+}
+
+struct NetworkWithNetnsInner {
+    netns_path: String,
+    entity_list: Vec<NetworkEntity>,
+}
+
+impl NetworkWithNetnsInner {
+    async fn new(config: &NetworkWithNetNsConfig) -> Result<Self> {
+        let entity_list = if config.netns_path.is_empty() {
+            warn!(sl!(), "skip to scan for empty netns");
+            vec![]
+        } else {
+            // get endpoint
+            get_entity_from_netns(config)
+                .await
+                .context("get entity from netns")?
+        };
+        Ok(Self {
+            netns_path: config.netns_path.to_string(),
+            entity_list,
+        })
+    }
+}
+
+pub(crate) struct NetworkWithNetns {
+    inner: Arc<RwLock<NetworkWithNetnsInner>>,
+}
+
+impl NetworkWithNetns {
+    pub(crate) async fn new(config: &NetworkWithNetNsConfig) -> Result<Self> {
+        Ok(Self {
+            inner: Arc::new(RwLock::new(NetworkWithNetnsInner::new(config).await?)),
+        })
+    }
+}
+
+#[async_trait]
+impl Network for NetworkWithNetns {
+    async fn setup(&self, h: &dyn Hypervisor) -> Result<()> {
+        let inner = self.inner.read().await;
+        let _netns_guard = netns::NetnsGuard::new(&inner.netns_path)
+            .context("net netns guard")
+            .unwrap();
+        for e in &inner.entity_list {
+            info!(sl!(), "attach endpoint start {}", e.endpoint.name().await);
+            e.endpoint.attach(h).await.context("attach")?;
+            info!(sl!(), "attach endpoint end {}", e.endpoint.name().await);
```

Comment:
I'm not entirely convinced we need all the `info!()` calls. As long as the returned error contains all relevant details, we could potentially remove some/all of these calls, or maybe just make them `debug!()` instead?

Reply:
Thanks for your review, remove the `info!()`.